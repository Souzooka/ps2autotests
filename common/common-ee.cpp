#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
