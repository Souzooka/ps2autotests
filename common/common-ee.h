#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
