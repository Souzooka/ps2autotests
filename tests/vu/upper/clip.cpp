#include <common-ee.h>
#include "../test_runner.h"

static const u32 __attribute__((aligned(16))) C_FULLPOS[4]  = {0x40000000, 0x40000000, 0x40000000, 0x40000000};
static const u32 __attribute__((aligned(16))) C_FULLNEG[4]  = {0xC0000000, 0xC0000000, 0xC0000000, 0xC0000000};
static const u32 __attribute__((aligned(16))) C_HALFPOS[4]  = {0x40000000, 0xC0000000, 0x40000000, 0xC0000000};
static const u32 __attribute__((aligned(16))) C_HALFNEG[4]  = {0xC0000000, 0x40000000, 0xC0000000, 0x40000000};
static const u32 __attribute__((aligned(16))) C_WZERO[4]    = {0x00000000, 0x00000000, 0x00000000, 0x00000000};
static const u32 __attribute__((aligned(16))) C_WNEGZERO[4] = {0x80000000, 0x80000000, 0x80000000, 0x80000000};
static const u32 __attribute__((aligned(16))) C_WNEGONE[4]  = {0xBF800000, 0xBF800000, 0xBF800000, 0xBF800000};
static const u32 __attribute__((aligned(16))) C_MAX[4]      = {0x7FFFFFFF, 0x7FFFFFFF, 0x7FFFFFFF, 0x7FFFFFFF};
static const u32 __attribute__((aligned(16))) C_MIN[4]      = {0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF};

static u32 *const CVF_FULLPOS  = (u32 *)(vu0_mem + 0x0070);
static u32 *const CVF_FULLNEG  = (u32 *)(vu0_mem + 0x0080);
static u32 *const CVF_HALFPOS  = (u32 *)(vu0_mem + 0x0090);
static u32 *const CVF_HALFNEG  = (u32 *)(vu0_mem + 0x00A0);
static u32 *const CVF_WZERO    = (u32 *)(vu0_mem + 0x00B0);
static u32 *const CVF_WNEGZERO = (u32 *)(vu0_mem + 0x00C0);
static u32 *const CVF_WNEGONE  = (u32 *)(vu0_mem + 0x00E0);
static u32 *const CVF_MAX      = (u32 *)(vu0_mem + 0x00F0);
static u32 *const CVF_MIN      = (u32 *)(vu0_mem + 0x0100);

static void setup_vf_constants() {
	*(vu128 *)CVF_FULLPOS  = *(vu128 *)C_FULLPOS;
	*(vu128 *)CVF_FULLNEG  = *(vu128 *)C_FULLNEG;
	*(vu128 *)CVF_HALFPOS  = *(vu128 *)C_HALFPOS;
	*(vu128 *)CVF_HALFNEG  = *(vu128 *)C_HALFNEG;
	*(vu128 *)CVF_WZERO    = *(vu128 *)C_WZERO;
	*(vu128 *)CVF_WNEGZERO = *(vu128 *)C_WNEGZERO;
	*(vu128 *)CVF_WNEGONE  = *(vu128 *)C_WNEGONE;
	*(vu128 *)CVF_MAX      = *(vu128 *)C_MAX;
	*(vu128 *)CVF_MIN      = *(vu128 *)C_MIN;
}

class ClipTestRunner : public TestRunner {
public:
	ClipTestRunner(int vu) : TestRunner(vu) {
	}

	void PerformBasic(const char* name, const u32 *s, const u32* t) {
		using namespace VU;

		printf("Basic test: %s -> ", name);

		Reset();

		WrLoadFloatRegister(DEST_XYZW, VF01, s);
		WrLoadFloatRegister(DEST_XYZW, VF02, t);
		
		Wr(CLIP(VF01, VF02));
		Wr(NOP());
		Wr(NOP());
		Wr(NOP());
		Wr(FCGET(VI01));
		
		Execute();

		PrintRegister(VI01, true);
	}
	
	void PerformMulti() {
		using namespace VU;

		printf("Multi test:\n");

		Reset();

		WrLoadFloatRegister(DEST_XYZW, VF01, CVF_FULLPOS);
		WrLoadFloatRegister(DEST_XYZW, VF02, CVF_FULLNEG);
		WrLoadFloatRegister(DEST_XYZW, VF03, CVF_HALFPOS);
		WrLoadFloatRegister(DEST_XYZW, VF04, CVF_HALFNEG);
		
		Wr(CLIP(VF01, VF00), FCGET(VI01));
		Wr(CLIP(VF02, VF00), FCGET(VI02));
		Wr(CLIP(VF03, VF00), FCGET(VI03));
		Wr(CLIP(VF04, VF00), FCGET(VI04));
		Wr(FCGET(VI05));
		Wr(FCGET(VI06));
		Wr(FCGET(VI07));
		Wr(FCGET(VI08));
		Wr(FCGET(VI09));
		Wr(FCGET(VI10));
		
		Execute();

		PrintRegister(VI01, true);
		PrintRegister(VI02, true);
		PrintRegister(VI03, true);
		PrintRegister(VI04, true);
		PrintRegister(VI05, true);
		PrintRegister(VI06, true);
		PrintRegister(VI07, true);
		PrintRegister(VI08, true);
		PrintRegister(VI09, true);
		PrintRegister(VI10, true);
	}
	
	void PerformMulti2() {
		using namespace VU;

		printf("Multi test 2:\n");

		Reset();

		WrLoadFloatRegister(DEST_XYZW, VF01, CVF_FULLPOS);
		WrLoadFloatRegister(DEST_XYZW, VF02, CVF_FULLNEG);
		WrLoadFloatRegister(DEST_XYZW, VF03, CVF_HALFPOS);
		WrLoadFloatRegister(DEST_XYZW, VF04, CVF_HALFNEG);
		WrLoadFloatRegister(DEST_XYZW, VF05, CVF_WNEGONE);
		
		Wr(CLIP(VF01, VF00));
		Wr(CLIP(VF01, VF05));
		Wr(CLIP(VF02, VF00));
		Wr(CLIP(VF02, VF05));
		Wr(CLIP(VF03, VF00));
		Wr(CLIP(VF03, VF05), FCGET(VI01));
		Wr(CLIP(VF04, VF00));
		Wr(CLIP(VF04, VF05), FCGET(VI02));
		Wr(NOP());
		Wr(FCGET(VI03));
		Wr(NOP());
		Wr(FCGET(VI04));
		
		Execute();

		PrintRegister(VI01, true);
		PrintRegister(VI02, true);
		PrintRegister(VI03, true);
		PrintRegister(VI04, true);
	}
};

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");
	
	setup_vf_constants();
	
	ClipTestRunner runner(0);
	
	runner.PerformBasic("FULL POS x -1", CVF_FULLPOS, CVF_WNEGONE);
	runner.PerformBasic("FULL NEG x -1", CVF_FULLNEG, CVF_WNEGONE);
	runner.PerformBasic("HALF POS x -1", CVF_HALFPOS, CVF_WNEGONE);
	runner.PerformBasic("HALF NEG x -1", CVF_HALFNEG, CVF_WNEGONE);
	runner.PerformBasic("HALF NEG x  0", CVF_HALFNEG, CVF_WZERO);
	runner.PerformBasic("HALF NEG x -0", CVF_HALFNEG, CVF_WNEGZERO);
	runner.PerformBasic(" 0 x -0", CVF_WZERO, CVF_WNEGZERO);
	runner.PerformBasic("-0 x  0", CVF_WNEGZERO, CVF_WZERO);
	runner.PerformBasic("FULL POS x FULL POS", CVF_FULLPOS, CVF_FULLPOS);
	runner.PerformBasic("HALF NEG x HALF NEG", CVF_HALFNEG, CVF_HALFNEG);
	runner.PerformBasic("MIN x 0", CVF_MIN, CVF_WZERO);
	runner.PerformBasic("MAX x 0", CVF_MAX, CVF_WZERO);
	runner.PerformBasic("0 x MIN", CVF_WZERO, CVF_MIN);
	runner.PerformBasic("0 x MAX", CVF_WZERO, CVF_MAX);
	runner.PerformMulti();
	runner.PerformMulti2();
	
	printf("-- TEST END\n");
	return 0;
}
