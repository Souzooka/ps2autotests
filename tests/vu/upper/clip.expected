-- TEST BEGIN
Basic test: FULL POS x -1 -> 0015
Basic test: FULL NEG x -1 -> 002a
Basic test: HALF POS x -1 -> 0019
Basic test: HALF NEG x -1 -> 0026
Basic test: HALF NEG x  0 -> 0026
Basic test: HALF NEG x -0 -> 0026
Basic test:  0 x -0 -> 0000
Basic test: -0 x  0 -> 0000
Basic test: FULL POS x FULL POS -> 0000
Basic test: HALF NEG x HALF NEG -> 0000
Basic test: MIN x 0 -> 002a
Basic test: MAX x 0 -> 0015
Basic test: 0 x MIN -> 0000
Basic test: 0 x MAX -> 0000
Multi test:
0000
0000
0000
0000
0015
056a
0a99
0666
0666
0666
Multi test 2:
0555
0aaa
0659
09a6
-- TEST END
